### PR TITLE
Add a basic manifest repository implementation

### PR DESCRIPTION
**Pull request by MoisesGSalas**
## Description

Implement `BaseManifest` to render a drydock environment.

A manifest repository is a collection of configuration files that describe that describe the state of our applications. This collection is meant to be tracked by a version control tool and enable a GitOps framework.

The `BaseManifest` implementation serves as a basic approach to this for Kubernetes. It mixes Kustomize and a Tutor (v13.3.1) environment to render a new environment that uses the Tutor env as a base and adds additional extensions as new resources. A drydock environment generated using `BaseManifest` looks like this:

```
drydockenv
├── base
│   ├── apps
│   ├── k8s
│   ├── kustomization.yml
│   └── plugins
├── extensions
│   ├── flowers.yml
│   ├── hpa.yml
│   ├── ingress.yml
│   ├── kustomization.yml
│   └── overrides.yml
└── kustomization.yml
```

Where `base` is standard Tutor env (minus files not needed by k8s), and extension is an additional Kustomize overlay with extra resources or patches.

The drydock environment is generated by using the tutor renderer on the templates defined on `default`. The templates on `default/base` are taken almost verbatim from the latest tutor version for the Maple release (v13.3.1), the only difference being the option to skip jobs (The current implementation of Jobs on Tutor doesn't play nicely with GitOps).

```diff
# Kustomization.yml
 resources:
 - k8s/namespace.yml
 - k8s/deployments.yml
-- k8s/jobs.yml
 - k8s/services.yml
 - k8s/volumes.yml
+{%- if not DRYDOCK_DISABLE_JOBS | default(false) %}
+- k8s/jobs.yml
+{%- endif %}
+

```

The `default/extension` includes additional resources such as an ingress, an Horizontal Pod Autoscaler and a flowers deployment. It also includes an `overrides.yml` (inspired by [this commit](https://github.com/overhangio/tutor/commit/b8f773a5efde67a769301aecf8434b8c19bbea93)) to modify the values of the resources in `default/base`.

## Additional info

- `BaseManifest` is static in the sense that is tied to the templates found in `default` and those are locked to the ones found in tutor v13.3.1 for the most part. In the future it may be possible to have different versions of `default`.
- Tutor is not exactly meant to be used as a library at the moment and the current implementation doesn't leverage the full tooling that it provides. At the moment we render a full tutor env (including our cherry-picked templates) and selectively move the ones we are interested.
- We are not making checks for conflicting options (e.g. `ENABLE_WEB_PROXY: true` and `DRYDOCK_INGRESS: true`)

### CODE REVIEW

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        config: DrydockConfig
+            A Tutor configuration extension.
+        """
+        tutor_root = config.get_root()
+        dst = path_join(tutor_root, self.options.get("output", "drydock-env"))
+        with tempfile.TemporaryDirectory() as tmpdir:
+            src = path_join(tmpdir)
+            self.render(src, config)
+            shutil.rmtree(path=dst, ignore_errors=True)
+            self.relocate_env(src, dst)
```
Comment (mariajgrimaldi) [inline drydock/manifest_builder/infrastructure/tutor_based_manifests.py:91]: This is my current folder structure:
```
├── config.yml
├── env
│   ├── base
│   │   ├── apps
│   │   │   ├── caddy
│   │   │   ├── openedx
│   │   │   └── redis
│   │   ├── k8s
│   │   │   ├── deployments.yml
│   │   │   ├── jobs.yml
│   │   │   ├── namespace.yml
│   │   │   ├── services.yml
│   │   │   └── volumes.yml
│   │   ├── kustomization.yml
│   │   └── plugins
│   │       ├── forum
│   │       ├── notes
│   │       └── settingmanager.yml
│   ├── extensions
│   │   ├── flowers.yml
│   │   ├── hpa.yml
│   │   ├── ingress.yml
│   │   ├── kustomization.yml
│   │   └── overrides.yml
│   └── kustomization.yml
├── reference.yml
└── requirements.txt
```
My tutor plugins list:
```
android         	(disabled)	13.0.0
discovery       	(disabled)	13.0.1
distro          	          	2.1.0
drydock         	          	0.1.0
ecommerce       	(disabled)	13.0.1
forum           	          	13.0.0
license         	(disabled)	13.0.0
mfe             	(disabled)	13.0.6
minio           	          	13.0.1
notes           	          	13.0.1
richie          	(disabled)	13.0.2
settingmanager  	          	0.1.0
webui           	(disabled)	13.0.2
xqueue          	(disabled)	13.0.0
```
Then I run: 
`tutor drydock save -r reference.yml`
When the command finishes, I run:
`tutor plugins list`
And this happens:
```
⚠️  Failed to enable plugin 'settingmanager': plugin 'settingmanager' is not installed.
android    	(disabled)	13.0.0
discovery  	(disabled)	13.0.1
distro     	          	2.1.0
drydock    	          	0.1.0
ecommerce  	(disabled)	13.0.1
forum      	          	13.0.0
license    	(disabled)	13.0.0
mfe        	(disabled)	13.0.6
minio      	          	13.0.1
notes      	          	13.0.1
richie     	(disabled)	13.0.2
webui      	(disabled)	13.0.2
xqueue     	(disabled)	13.0.0
```
Is that okay?

File: drydock/manifest_builder/infrastructure/tutor_config.py [hunk 2]
```diff
@@ -1,14 +1,61 @@
...
+            "DRYDOCK_LMS_WORKER_REQUEST_MEMORY": "1Gi",
+            "DRYDOCK_CMS_LIMIT_CPU": 1,
+            "DRYDOCK_CMS_LIMIT_MEMORY": "2Gi",
+            "DRYDOCK_CMS_REQUEST_CPU": "600m",
+            "DRYDOCK_CMS_REQUEST_MEMORY": "1Gi",
+            "DRYDOCK_CMS_WORKER_LIMIT_CPU": 1,
+            "DRYDOCK_CMS_WORKER_LIMIT_MEMORY": "2Gi",
+            "DRYDOCK_CMS_WORKER_REQUEST_CPU": "600m",
+            "DRYDOCK_CMS_WORKER_REQUEST_MEMORY": "1Gi",
+        }
```
Comment (mariajgrimaldi) [inline drydock/manifest_builder/infrastructure/tutor_config.py:37]: self.defaults changes? why this dictionary is not constant to the class?

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        fmt.echo_info(f"Drydock environment generated in {dst}")
```
Comment (mariajgrimaldi) [inline drydock/manifest_builder/infrastructure/tutor_based_manifests.py:93]: is it relevant to print that the env was generated in a temp environment?
```
Environment generated in /tmp/tmp9ibnztrx/env
Drydock environment generated in /home/mgmdi/work/atlas/manifests-admiral/env
```

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        Parameters
+        ----------
+        src: str
+            The inital path where the full tutor env was rendered.
+        dst: str
+            The path to save the final drydock env.
+        """
+        base_dir = path_join(src, "env/drydock/default")
+        plugins_dir = path_join(src, "env/plugins")
```
Comment (mariajgrimaldi) [inline drydock/manifest_builder/infrastructure/tutor_based_manifests.py:68]: Can we use these as constants?

**Comment by mariajgrimaldi** [comment]
How did you decide what to add as configurable in the base templates?

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py (hunk 1 above)
Comment (MoisesGSalas) [reply drydock/manifest_builder/infrastructure/tutor_based_manifests.py:91]: Yes (actually is not ok, but is a current limitation). At the moment we don't render module nor yaml plugins on `base/plugins`. The current workaround is to define the `TUTOR_PLUGINS_ROOTS` outside `env`.

The folder structure would look something like this:

```
├── config.yml
├── env
│   ├── base
│   │   ├── apps
│   │   │   ├── caddy
│   │   │   ├── openedx
│   │   │   └── redis
│   │   ├── k8s
│   │   │   ├── deployments.yml
│   │   │   ├── jobs.yml
│   │   │   ├── namespace.yml
│   │   │   ├── services.yml
│   │   │   └── volumes.yml
│   │   ├── kustomization.yml
│   │   └── plugins
│   │       ├── forum
│   │       └──  notes
│   ├── extensions
│   │   ├── flowers.yml
│   │   ├── hpa.yml
│   │   ├── ingress.yml
│   │   ├── kustomization.yml
│   │   └── overrides.yml
│   └── kustomization.yml
├── reference.yml
└── requirements.txt
└── plugins
    └── settingmanager.yml
```

and your `TUTOR_PLUGINS_ROOT` should be something like `$TUTOR_ROOT/plugins`

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py (hunk 3 above)
Comment (MoisesGSalas) [reply drydock/manifest_builder/infrastructure/tutor_based_manifests.py:93]: It's not, but that comes from the tutor [`env.save()`](https://github.com/overhangio/tutor/blob/698da4b1adfa9509876a9bd6b09cb755f87b137e/tutor/env.py#L253-L265) function. The second line is what drydock prints.

**Comment by MoisesGSalas** [comment]
> How did you decide what to add as configurable in the base templates?

The ones in base I tried to keep almost the same as the ones in [tutor](https://github.com/overhangio/tutor/tree/v13.3.1/tutor/templates). Minus the Jobs for reasons explained above.

The ones in extensions I've opted for making all the additional resources toggleable

File: drydock/manifest_builder/infrastructure/tutor_config.py (hunk 2 above)
Comment (MoisesGSalas) [reply drydock/manifest_builder/infrastructure/tutor_config.py]: I've opted to define the defaults on a `defaults.yml` on each template set and load them from there.

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py (hunk 4 above)
Comment (MoisesGSalas) [reply drydock/manifest_builder/infrastructure/tutor_based_manifests.py]: Done

File: drydock/manifest_builder/infrastructure/tutor_based_manifests.py (hunk 3 above)
Comment (mariajgrimaldi) [reply drydock/manifest_builder/infrastructure/tutor_based_manifests.py:97]: Right! :thinking:

**Review by mariajgrimaldi — APPROVED** [review]
This looks good! I tried it, and it seems to be working as expected. Just one more thing: you could add a short how-to describing how to generate manifests & how to apply them.

**Review by jfavellar90 — APPROVED** [review]

**Pull request merged**